### PR TITLE
DEPR: Deprecate dtype inference on pandas objects

### LINKED ISSUE #56012: DEPR: Series and Index shouldn't do inference on pandas objects
The following changes the dtype, which is unexpected:

```
idx = pd.Index(pd.date_range("2015-01-01 10:00", freq="D", periods=3), dtype=object)
pd.Index(idx).dtype
```

this returns ``datetime64[ns]``

Doing inference on a pandas object is pretty weird, this makes stuff very complicated with the new string option. Set_Index can change the dtype, e.g.

```
df = pd.DataFrame({"a": pd.Series(pd.date_range("2015-01-01 10:00", freq="D", periods=3), dtype=object), "b": 1})
df.dtypes

a    object
b     int64
dtype: object


df.set_index("a").index.dtype
datetime64[ns]
```

Users can just call infer_objects if they want to do inference. We should keep the inference logic for non-pandas objects

cc @jbrockmendel

### PR DESCRIPTION
**Pull request by phofl**
- [ ] closes #56012 (Replace xxxx with the GitHub issue number)
- [ ] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [ ] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [ ] Added [type annotations](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#type-hints) to new arguments/methods/functions.
- [ ] Added an entry in the latest `doc/source/whatsnew/vX.X.X.rst` file if fixing a bug or adding a new feature.

This is very weird to begin with and becomes a real PITA when we infer strings as arrow backed strings

### CODE REVIEW

**Comment by phofl** [comment]
There was some weird additional inference going on in cat, which was uncovered by the deprecation, is fixed now but doesn't really need a whatsnew since we merged another fix today as well

**Comment by phofl** [comment]
cc @mroeschke 

good to merge here?

File: pandas/core/indexes/base.py [hunk 1]
```diff
@@ -571,7 +573,18 @@ def __new__(
...
         arr = klass._ensure_array(arr, arr.dtype, copy=False)
-        return klass._simple_new(arr, name, refs=refs)
+        result = klass._simple_new(arr, name, refs=refs)
+        if dtype is None and is_pandas_object and data_dtype == np.object_:
+            if result.dtype != data_dtype:
+                warnings.warn(
+                    "Dtype inference on a pandas object "
+                    "(Series, Index, ExtensionArray) is deprecated. The Index "
+                    "constructor will keep the original dtype in the future. "
+                    "Call ``infer_objects on the result",
```
Comment (mroeschke) [inline pandas/core/indexes/base.py:583]:
```suggestion
                    "Call `infer_objects` on the result",
```

File: pandas/core/frame.py [hunk 2]
```diff
@@ -912,6 +916,17 @@ def __init__(
 
         NDFrame.__init__(self, mgr)
 
+        if original_dtype is None and is_pandas_object and data_dtype == np.object_:
+            if self.dtypes.iloc[0] != data_dtype:
+                warnings.warn(
+                    "Dtype inference on a pandas object "
+                    "(Series, Index, ExtensionArray) is deprecated. The Index "
+                    "constructor will keep the original dtype in the future. "
+                    "Call ``infer_objects on the result",
```
Comment (mroeschke) [inline pandas/core/frame.py:925]:
```suggestion
                    "Call `infer_objects` on the result",
```

File: pandas/core/series.py [hunk 3]
```diff
@@ -581,6 +585,17 @@ def __init__(
         self.name = name
         self._set_axis(0, index)
 
+        if original_dtype is None and is_pandas_object and data_dtype == np.object_:
+            if self.dtype != data_dtype:
+                warnings.warn(
+                    "Dtype inference on a pandas object "
+                    "(Series, Index, ExtensionArray) is deprecated. The Index "
+                    "constructor will keep the original dtype in the future. "
+                    "Call ``infer_objects on the result",
```
Comment (mroeschke) [inline pandas/core/series.py:594]:
```suggestion
                    "Call `infer_objects` on the result",
```

File: pandas/core/series.py [hunk 4]
```diff
@@ -581,6 +585,17 @@ def __init__(
...
+                    FutureWarning,
+                    stacklevel=2,
```
Comment (mroeschke) [inline pandas/core/series.py:596]: Does `find_stack_level` return the incorrect result here?

Comment (phofl) [reply pandas/core/series.py:596]: This is nice for internal debugging, but yeah I can use find_stack_level as well

Comment (phofl) [reply pandas/core/series.py]: updated

File: pandas/_testing/__init__.py [hunk 5]
```diff
@@ -285,11 +286,17 @@ def box_expected(expected, box_cls, transpose: bool = True):
         else:
             expected = pd.array(expected, copy=False)
     elif box_cls is Index:
-        expected = Index(expected)
+        with warnings.catch_warnings():
+            warnings.filterwarnings("ignore", "Dtype inference", category=FutureWarning)
```
Comment (jbrockmendel) [inline pandas/_testing/__init__.py:290]: these warnings seem like they might add up in test runtime.  might be more performant to pass dtype=expected.dtype to the constructors?  (i guess that is making a not-enforced assumption about what expected is?)

File: pandas/core/frame.py [hunk 6]
```diff
@@ -912,6 +916,17 @@ def __init__(
 
         NDFrame.__init__(self, mgr)
 
+        if original_dtype is None and is_pandas_object and data_dtype == np.object_:
+            if self.dtypes.iloc[0] != data_dtype:
+                warnings.warn(
+                    "Dtype inference on a pandas object "
+                    "(Series, Index, ExtensionArray) is deprecated. The Index "
+                    "constructor will keep the original dtype in the future. "
+                    "Call `infer_objects` on the result",
```
Comment (jbrockmendel) [inline pandas/core/frame.py:925]: "to get the old behavior"

File: pandas/core/frame.py [hunk 7]
```diff
@@ -912,6 +916,17 @@ def __init__(
 
         NDFrame.__init__(self, mgr)
 
+        if original_dtype is None and is_pandas_object and data_dtype == np.object_:
+            if self.dtypes.iloc[0] != data_dtype:
+                warnings.warn(
+                    "Dtype inference on a pandas object "
```
Comment (jbrockmendel) [inline pandas/core/frame.py:922]: "in the DataFrame constructor" somewhere in this sentence?  I think the "The Index constructor" below is a copy/paste leftover

File: pandas/core/frame.py [hunk 8]
```diff
@@ -912,6 +916,17 @@ def __init__(
 
         NDFrame.__init__(self, mgr)
 
+        if original_dtype is None and is_pandas_object and data_dtype == np.object_:
```
Comment (jbrockmendel) [inline pandas/core/frame.py:919]: do you want to do anything with dict_of_pandas_objects?

**Comment by jbrockmendel** [comment]
Should this affect pd.array?

File: pandas/_testing/__init__.py (hunk 5 above)
Comment (phofl) [reply pandas/_testing/__init__.py:290]: we sometimes rely on the inference behaviour, that's why I kept this as is for now, I don't want to screw up testing before we enforce the deprecation

File: pandas/core/frame.py (hunk 6 above)
Comment (phofl) [reply pandas/core/frame.py:925]: added

File: pandas/core/frame.py (hunk 7 above)
Comment (phofl) [reply pandas/core/frame.py:922]: yeah sorry, adjusted

File: pandas/core/frame.py (hunk 8 above)
Comment (phofl) [reply pandas/core/frame.py:919]: That is the funny thing, they already preserve the dtype...

**Comment by phofl** [comment]
> Should this affect pd.array?

Yes, but it already preserves the dtype, so we should be good there

File: doc/source/whatsnew/v2.2.0.rst [hunk 9]
```diff
@@ -460,6 +460,7 @@ Other Deprecations
 - Deprecated behavior of :meth:`Index.insert` with an object-dtype index silently performing type inference on the result, explicitly call ``result.infer_objects(copy=False)`` for the old behavior instead (:issue:`51363`)
 - Deprecated casting non-datetimelike values (mainly strings) in :meth:`Series.isin` and :meth:`Index.isin` with ``datetime64``, ``timedelta64``, and :class:`PeriodDtype` dtypes (:issue:`53111`)
 - Deprecated downcasting behavior in :meth:`Series.where`, :meth:`DataFrame.where`, :meth:`Series.mask`, :meth:`DataFrame.mask`, :meth:`Series.clip`, :meth:`DataFrame.clip`; in a future version these will not infer object-dtype columns to non-object dtype, or all-round floats to integer dtype. Call ``result.infer_objects(copy=False)`` on the result for object inference, or explicitly cast floats to ints. To opt in to the future version, use ``pd.set_option("future.no_silent_downcasting", True)`` (:issue:`53656`)
+- Deprecated dtype inference in :class:`Index`, :class:`Series` and :class:`DataFrame` constructors when giving a pandas input, call ``.infer_objects`` on the input to keep the current behavior (:issue:`56012`)
```
Comment (jbrockmendel) [inline doc/source/whatsnew/v2.2.0.rst:463]: on the input -> on the result?

Comment (phofl) [reply doc/source/whatsnew/v2.2.0.rst:463]: either was is totally fine

File: pandas/tests/series/methods/test_between.py [hunk 10]
```diff
@@ -20,7 +20,7 @@ def test_between(self):
         tm.assert_series_equal(result, expected)
 
     def test_between_datetime_object_dtype(self):
-        ser = Series(bdate_range("1/1/2000", periods=20).astype(object))
+        ser = Series(bdate_range("1/1/2000", periods=20).astype(object), dtype=object)
```
Comment (jbrockmendel) [inline pandas/tests/series/methods/test_between.py:23]: The .astype(object) here is redundant

File: pandas/tests/series/accessors/test_dt_accessor.py [hunk 11]
```diff
@@ -259,9 +259,10 @@ def test_dt_accessor_limited_display_api(self):
         tm.assert_almost_equal(results, sorted(set(ok_for_dt + ok_for_dt_methods)))
 
         # Period
-        ser = Series(
-            period_range("20130101", periods=5, freq="D", name="xxx").astype(object)
-        )
+        with tm.assert_produces_warning(FutureWarning, match="Dtype inference"):
+            ser = Series(
+                period_range("20130101", periods=5, freq="D", name="xxx").astype(object)
```
Comment (jbrockmendel) [inline pandas/tests/series/accessors/test_dt_accessor.py:264]: can you construct the index outside the context so it is obvious where the warning comes from

File: pandas/tests/series/methods/test_between.py (hunk 10 above)
Comment (phofl) [reply pandas/tests/series/methods/test_between.py:23]: ah yes, changed

File: pandas/tests/series/accessors/test_dt_accessor.py (hunk 11 above)
Comment (phofl) [reply pandas/tests/series/accessors/test_dt_accessor.py:264]: moved

**Review by mroeschke — APPROVED** [review]

**Pull request merged**